### PR TITLE
Add SecureSourceManagerRepository controller and mockgcp

### PR DESCRIPTION
**Pull request by ericpang777**
### Change description

Fixes #

### Tests you have done

- [x] Run `make ready-pr` to ensure this PR is ready for review.
- [x] Perform necessary E2E testing for changed resources.

### CODE REVIEW

**Comment by ericpang777** [comment]
@gemmahou Chain PR off https://github.com/GoogleCloudPlatform/k8s-config-connector/pull/3120

File: pkg/test/resourcefixture/testdata/basic/securesourcemanager/securesourcemanagerrepository/securesourcemanagerrepositorybasic/create.yaml [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+apiVersion: securesourcemanager.cnrm.cloud.google.com/v1alpha1
+kind: SecureSourceManagerRepository
+metadata:
+  name: ssmrepository-${uniqueId}
+  annotations:
+      alpha.cnrm.cloud.google.com/reconciler: "direct"
```
Comment (gemmahou) [inline pkg/test/resourcefixture/testdata/basic/securesourcemanager/securesourcemanagerrepository/securesourcemanagerrepositorybasic/create.yaml:20]: For new resources, we do not need this annotation. It's used to opt-in direct reconciliation for resources that using TF/DCL controller.

File: pkg/test/resourcefixture/testdata/basic/securesourcemanager/securesourcemanagerrepository/securesourcemanagerrepositorybasic/create.yaml
Comment (gemmahou) [inline pkg/test/resourcefixture/testdata/basic/securesourcemanager/securesourcemanagerrepository/securesourcemanagerrepositorybasic/create.yaml:1]: Do we need update.yaml? If not, could you explain a bit?

We also need a "full" test to cover required and optional fields. That's "Beta standard".

File: pkg/test/resourcefixture/testdata/basic/securesourcemanager/securesourcemanagerrepository/securesourcemanagerrepositorybasic/create.yaml (hunk 1 above)
Comment (ericpang777) [reply pkg/test/resourcefixture/testdata/basic/securesourcemanager/securesourcemanagerrepository/securesourcemanagerrepositorybasic/create.yaml:20]: Done

File: pkg/test/resourcefixture/testdata/basic/securesourcemanager/securesourcemanagerrepository/securesourcemanagerrepositorybasic/create.yaml
Comment (ericpang777) [reply pkg/test/resourcefixture/testdata/basic/securesourcemanager/securesourcemanagerrepository/securesourcemanagerrepositorybasic/create.yaml:1]: We don't currently support Update for Repository. I mentioned this in the tracker doc, is that good enough?

Comment (ericpang777) [reply pkg/test/resourcefixture/testdata/basic/securesourcemanager/securesourcemanagerrepository/securesourcemanagerrepositorybasic/create.yaml:1]: Added a full test

File: pkg/controller/direct/securesourcemanager/repository_controller.go [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+	}
+	status.ExternalRef = &a.id.External
+	return createOp.UpdateStatus(ctx, status, nil)
+}
+
+func (a *SecureSourceManagerRepositoryAdapter) Update(ctx context.Context, updateOp *directbase.UpdateOperation) error {
+	log := klog.FromContext(ctx)
+	log.Info("update of SecureSourceManagerRepository not supported")
+	return nil
+}
```
Comment (gemmahou) [inline pkg/controller/direct/securesourcemanager/repository_controller.go:177]: Want to double check on this. Are `instance` and `instanceConfig` fields immutable? Can they be updated in the future?

**Comment by gemmahou** [comment]
Can we run the two tests you added against realGCP and upload the diff file between mockGCP and real? Want to take a look at if the diff is acceptable.

File: pkg/controller/direct/securesourcemanager/repository_controller.go (hunk 2 above)
Comment (ericpang777) [reply pkg/controller/direct/securesourcemanager/repository_controller.go:177]: There's no `instanceConfig` field. (Maybe you thought of the `initialConfig` field?)

For the `instance` field, no near-term plans to make it updatable. The `initialConfig` field is an input only field.

**Comment by ericpang777** [comment]
> Can we run the two tests you added against realGCP and upload the diff file between mockGCP and real? Want to take a look at if the diff is acceptable.

Running

**Comment by ericpang777** [comment]
> Can we run the two tests you added against realGCP and upload the diff file between mockGCP and real? Want to take a look at if the diff is acceptable.

Added a commit for the realgcp log

**Comment by ericpang777** [comment]
@gemmahou  I tried to make a more accurate mock controller, by removing the LRO integration (as our repository doesn't do that). However I noticed the removal of some fields in the golden.yaml. Is that okay? Should I switch my mock change back? 

The diff is in the latest commit 15a401bdf8295a7c97bf250fb8afa683d2df0957

**Comment by ericpang777** [comment]
The main difference between the previous mock and real logs was that the mock log is making requests to the Operations service, whereas the real one is not

File: pkg/controller/direct/securesourcemanager/repository_controller.go (hunk 2 above)
Comment (gemmahou) [reply pkg/controller/direct/securesourcemanager/repository_controller.go:177]: shall we mark `instance` and `initialConfig` immutable [here](https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/master/apis/securesourcemanager/v1alpha1/repository_types.go#L43-L49)?

Comment (gemmahou) [reply pkg/controller/direct/securesourcemanager/repository_controller.go:177]: Discussed offline, let's follow up on this in another PR!

**Comment by ericpang777** [comment]
> @gemmahou I tried to make a more accurate mock controller, by removing the LRO integration (as our repository doesn't do that). However I noticed the removal of some fields in the golden.yaml. Is that okay? Should I switch my mock change back?
> 
> The diff is in the latest commit [15a401b](https://github.com/GoogleCloudPlatform/k8s-config-connector/commit/15a401bdf8295a7c97bf250fb8afa683d2df0957)

Ok I just reverted the commit

File: pkg/controller/direct/securesourcemanager/repository_controller.go (hunk 2 above)
Comment (ericpang777) [reply pkg/controller/direct/securesourcemanager/repository_controller.go:177]: Will do!

**Comment by ericpang777** [comment]
@gemmahou Is the http log diff acceptable?

**Comment by gemmahou** [comment]
> 15a401b

The missing fields in golden object is strange and we need to understand why. I do not think it's related to the code change in this commit, but might be something wrong when you collect the log. 

    "message": "Resource \"projects/${projectId}/locations/us-central1/repositories/ssmrepository-${uniqueId}\" was not found",
    "message": "Resource 'projects/${projectId}/locations/us-central1/repositories/ssmrepository-${uniqueId}' was not found",
 This message diff should be a quick fix.
 
 "instance": "projects/${projectNumber}/locations/us-central1/instances/ssminstance-dep-${uniqueId}",
    "instance": "projects/${projectId}/locations/us-central1/instances/ssminstance-dep-${uniqueId}",
    
This instance diff we also need to double check because projectId and projectNumber are different

File: pkg/test/resourcefixture/testdata/basic/securesourcemanager/securesourcemanagerrepository/securesourcemanagerrepositorybasic/_http.log [hunk 3]
```diff
@@ -118,21 +118,17 @@ X-Xss-Protection: 0
 {
   "error": {
     "code": 404,
-    "message": "Resource 'projects/${projectId}/locations/us-central1/repositories/ssmrepository-${uniqueId}' was not found",
+    "message": "\"Resource 'projects/${projectId}/locations/us-central1/repositories/ssmrepository-${uniqueId}' was not found\"",
```
Comment (gemmahou) [inline pkg/test/resourcefixture/testdata/basic/securesourcemanager/securesourcemanagerrepository/securesourcemanagerrepositorybasic/_http.log:121]: nit: should be `"Resource \"projects/${projectId}/locations/us-central1/repositories/ssmrepository-${uniqueId}\" was not found"`

Comment (ericpang777) [reply pkg/test/resourcefixture/testdata/basic/securesourcemanager/securesourcemanagerrepository/securesourcemanagerrepositorybasic/_http.log]: Done oops

**Comment by gemmahou** [comment]
/lgtm

@yuwenma would you mind taking another look? this [commit](https://github.com/GoogleCloudPlatform/k8s-config-connector/pull/3156/commits/637636743333b8a46a216f4d0443b50460bc82af#diff-30fa6087f33144bc316e920dd674f9c74282b7ca24d50784e0a6af7cf46da88b) has the diff between realGCP and mock, I feel they are acceptable, most of them are related to LRO and http header.

**Comment by ericpang777** [comment]
test failure seems unrelated

**Comment by ericpang777** [comment]
@yuwenma Is the diff between the real and mock ok? It's really just the HTTP header and LRO fields that differ

File: pkg/controller/direct/securesourcemanager/repository_controller.go [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+	return gcpClient, err
+}
+
+func (m *modelSecureSourceManagerRepository) AdapterForObject(ctx context.Context, reader client.Reader, u *unstructured.Unstructured) (directbase.Adapter, error) {
+	obj := &krm.SecureSourceManagerRepository{}
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, &obj); err != nil {
+		return nil, fmt.Errorf("error converting to %T: %w", obj, err)
+	}
+
+	id, err := krm.NewSecureSourceManagerRepositoryRef(ctx, reader, obj)
```
Comment (yuwenma) [inline pkg/controller/direct/securesourcemanager/repository_controller.go:74]: @gemmahou just FYI This identity and reference need to be updated as well.

**Comment by yuwenma** [comment]
> @yuwenma Is the diff between the real and mock ok? It's really just the HTTP header and LRO fields that differ

Yes, this is pretty good.

**Pull request merged**